PDO refactor quoter hook to return zend_string*

Girgias opened the pull request:
This PR refactors the quoter hook to return a zend_string* instead of using out params as the drivers bundled in php-src always return true.

This in turn made some refactoring in the sql parser possible to also use zend_strings instead of a combination of `char*` and `size_t` elements. I've also added a couple more known strings for recurring values which can possibly be used outside of PDO.

nikic commented on Zend/zend_string.h:
```diff
@@ -551,6 +551,9 @@ EMPTY_SWITCH_DEFAULT_CASE()
 	_(ZEND_STR_MIXED,                  "mixed") \
+	_(ZEND_STR_0,                      "0") \
+	_(ZEND_STR_1,                      "1") \
+	_(ZEND_STR_QUESTION_MARK,          "?") \
```
You should use ZSTR_CHAR instead.

Girgias commented:
@adambaratz and @sim1984 could you test DBLIB and FireBird respectively to check I haven't messed it up? As we currently have no CI testing enabled for them.

adambaratz commented:
Thanks for flagging me! Still looking through this. Will try to wrap up my review today or tomorrow.

adambaratz reviewed (commented):
Found one compile error, but otherwise looks good! Tests pass and didn't uncover any memory leaks.

Girgias commented on ext/pdo_mysql/mysql_driver.c:
```diff
@@ -319,24 +322,28 @@ static int mysql_handle_quoter(pdo_dbh_t *dbh, const char *unquoted, size_t unqu
+
+	quoted_str = zend_string_init(quoted, quotedlen, 0);
+	efree(quoted);
+	return quoted_str;
```
Memo: drop this

nikic commented on ext/pdo/pdo_dbh.c:
```diff
@@ -1143,13 +1142,9 @@ PHP_METHOD(PDO, quote)
-	}
-	PDO_HANDLE_DBH_ERR();
-	RETURN_FALSE;
+	quoted_string = dbh->methods->quoter(dbh, str, paramtype);
```
You can inline this variable...

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -235,21 +233,17 @@ safe:
-							goto clean_up;
-						}
+
+						quoted_buf = stmt->dbh->methods->quoter(stmt->dbh, buf, param->param_type);
```
Same here, can inline the variable.

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -70,8 +70,7 @@ static int scan(Scanner *s)
-	size_t qlen;	/* quoted length of value */
-	char *quoted;	/* quoted value */
+	zend_string *quoted;	/* quoted value */
 	int freeq;
```
Do you need freeq still? You should be able to say that it should always be released (releasing interned strings does nothing).

Girgias replied:
Wasn't sure in regards to the behaviour with interned strings, but will do this.

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -270,40 +263,41 @@ safe:
+						default: {
+							zend_string *quoted_buf;
+
+							/* TODO Should this be zval_try_get_string_func() ? */
```
Just zval_try_get_string, but yes, that would be more elegant than the EG(exception) check.

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -270,40 +263,41 @@ safe:
 								/* bork */
 								ret = -1;
 								strncpy(stmt->error_code, stmt->dbh->error_code, 6);
+								/* TODO Is this dead code now? */
```
With zval_try_get_string at least, yes.

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -317,10 +311,9 @@ safe:
 				}
-				plc->quoted = Z_STRVAL_P(parameter);
-				plc->qlen = Z_STRLEN_P(parameter);
+				plc->quoted = Z_STR_P(parameter);
```
if freeq is dropped, this one would need a zend_string_copy.

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -386,24 +380,23 @@ rewrite:
 			/* check if bound parameter is already available */
 			if (!strcmp(name, "?") || (p = zend_hash_str_find_ptr(stmt->bound_param_map, name, plc->len)) == NULL) {
-				spprintf(&idxbuf, 0, tmpl, bind_no++);
+				idxbuf = strpprintf(0, tmpl, bind_no++);
```
```suggestion
				idxbuf = zend_strpprintf(0, tmpl, bind_no++);
```

nikic commented on ext/pdo/pdo_sql_parser.re:
```diff
@@ -386,24 +380,23 @@ rewrite:
 			if (!skip_map && stmt->named_rewrite_template) {
 				/* create a mapping */
-				zend_hash_str_update_mem(stmt->bound_param_map, name, plc->len, idxbuf, plc->qlen + 1);
+				zend_hash_str_update_mem(stmt->bound_param_map, name, plc->len, ZSTR_VAL(plc->quoted), ZSTR_LEN(plc->quoted) + 1);
```
It would be more elegant to switch bound_param_map to store zend_string* instead, which means you wouldn't need to create a separate copy of the string to store in this map (_mem performs a copy).

Girgias replied:
Will look into this

Girgias replied:
bound_param_map is a HashTable, so I'm not sure how one can go about to do that (I'm rather unfamiliar with the whole HashTable API and how it works under the hood).

nikic replied:
In this case, you would modify free_param_name to do a zend_string_release instead and use zend_hash_str_update_ptr instead of zend_hash_str_update_mem to insert the zend_string pointer.

nikic commented on ext/pdo_dblib/dblib_driver.c:
```diff
@@ -165,34 +165,36 @@ static int dblib_handle_quoter(pdo_dbh_t *dbh, const char *unquoted, size_t unqu
+		++quotedlen; /* N prefix */
 	}
-	q = *quoted = emalloc(*quotedlen + 1); /* Add byte for terminal null */
+	q = quoted = emalloc(quotedlen + 1); /* Add byte for terminal null */
```
Why not zend_string_alloc?

Girgias replied:
As I haven't tried to set up dblib nor firebird locally I went with the most minimal change, although not doing stuff on a zend_string is rather stupid I agree, will look into this.

nikic commented on ext/pdo_firebird/firebird_driver.c:
```diff
@@ -651,42 +651,42 @@ static zend_long firebird_handle_doer(pdo_dbh_t *dbh, const char *sql, size_t sq
-	*quotedlen = unquotedlen + qcount + 2;
-	*quoted = c = emalloc(*quotedlen+1);
+	quotedlen = ZSTR_LEN(unquoted) + qcount + 2;
+	quoted = c = emalloc(quotedlen+1);
```
Same here, why not directly zend_string_alloc?

nikic commented on ext/pdo_dblib/dblib_driver.c:
```diff
@@ -185,16 +186,12 @@ static zend_string* dblib_handle_quoter(pdo_dbh_t *dbh, const zend_string *unquo
-	*q++ = '\'';
-
-	*q = 0;
+	*q = '\'';
```
It's still necessary to null terminate the string.

nikic commented on ext/pdo_dblib/dblib_driver.c:
```diff
@@ -185,16 +186,12 @@ static zend_string* dblib_handle_quoter(pdo_dbh_t *dbh, const zend_string *unquo
 }
 
-static bool pdo_dblib_transaction_cmd(const char *cmd, pdo_dbh_t *dbh)
+static bool pdo_dblib_transaction_cmd(const char *qmd, pdo_dbh_t *dbh)
```
Overzealous search and replace?

Girgias commented:
Thanks for taking care of the bits I didn't manage.

I've squashed everything related into the first two commits and added an entry into UPGRADING.INTERNALS.

nikic commented:
It would be nice to switch the `doer` to use zend_string as well. It's less directly useful there, but it will make it consistent with preparer and quoter.

Girgias commented:
> It would be nice to switch the `doer` to use zend_string as well. It's less directly useful there, but it will make it consistent with preparer and quoter.

ACK
